src/System.Management.Automation/engine/parser/AstVisitor.cs

```diff
@@ -171,6 +171,10 @@ public interface ICustomAstVisitor2 : ICustomAstVisitor
 
         /// <summary/>
         object VisitDynamicKeywordStatement(DynamicKeywordStatementAst dynamicKeywordAst);
+        /// <summary/>
+        object VisitStatementChain(StatementChainAst statementChainAst);
+        /// <summary/>
+        object VisitPipelineChain(PipelineChainAst pipelineChainAst);
```

Comment:
Until default interface methods are added this will be a major breaking change.  Any assembly that has an `ICustomAstVisitor2` implementation will throw a `TypeLoadException` on load.

(I know this is still in draft and you may already plan to change this, just making sure)

Reply:
Yep planning to change this.

But the real question is to what.

I've implemented it like this to get it working, but I think this is the first new AST type added since 5 and we need to decide:

- Do we have an ICustomAstVisitor3? (And if so, will it be under mutation until PS 7 is GA?)
- Do I find a way to fold this AST into others?
- Is there some third option I haven't considered?

Reply:
BTW I'm hoping to hook up the last of the compilation today and then I'll update the description with all the discussions we need to have.

Basically I decided to implement this in lieu of an RFC not to get it merged but so we can talk concretely about a proposed implementation

Reply:
> I've implemented it like this to get it working, but I think this is the first new AST type added since 5 and we need to decide:

I think this is why tbh.  At least it's the reason I've strayed away from it personally.

> Do we have an ICustomAstVisitor3? (And if so, will it be under mutation until PS 7 is GA?)

For two ASTs probably not.  If there was a push to get ternary and null coalescing operators in for 7 though, I think there would be a stronger case.

> Do I find a way to fold this AST into others?

That's what I ended up doing for [property accessors](https://github.com/SeeminglyScience/PowerShell/blob/2f760a959a12edfd5cc38381dc8c2de8b98de5d7/src/System.Management.Automation/engine/parser/ast.cs#L3191-L3199).  Originally I wanted to make a `AccessorBodyAst` but I couldn't come up with a good way to handle this.

> Is there some third option I haven't considered?

Wait until default interface methods 😕. That's assuming the way they implement it allows for this scenario, I'm pretty sure it would but afaict it's still in an early prototyping phase.

Reply:
> If there was a push to get ternary and null coalescing operators in for 7 though

I believe we're aiming to do work on those -- hence why I think ICustomAstVisitor3 might actually be worth considering.

But I really don't like how complex that will make targeting multiple PowerShell versions. I'd say I wish `AstVisitor` was all we had, but really that just pushes breakage into what downstream developers must override...

Reply:
> But I really don't like how complex that will make targeting multiple PowerShell versions. I'd say I wish AstVisitor was all we had, but really that just pushes breakage into what downstream developers must override...

I'd like to see `ICustomAstVisitor*` deprecated and introduce two new visitors that somewhat follow what roslyn does:

```csharp
public class CustomAstVisitor
{
    public virtual void VisitScriptBlock(ScriptBlockAst scriptBlockAst) => DefaultVisit(scriptBlockAst);
    // etc
    public virtual void DefaultVisit(Ast ast) { }
}

public class CustomAstVisitor<TResult>
{
    public virtual TResult VisitScriptBlock(ScriptBlockAst scriptBlockAst) => DefaultVisit(scriptBlockAst);
    // etc
    public virtual TResult DefaultVisit(Ast ast) => default;
}
```

That way you can still *somewhat* handle new ASTs without specifically targeting the newest version.  It does eliminate the possibility of using a `struct` but they get boxed atm anyway.

If `ICustomAstVisitor*` was considered deprecated you skip adding this one to it.

Reply:
The default interface method is out in the .NET Core 3.0 preview 5 (the one powershell adopts). Though haven't tried it myself, it seems we can get away from the `ICustomAstVisitor3` problem.

Reply:
@daxian-dbw Oooo it sure is.  Just gave it a spin with a similar scenario, it solves the problem 🎉 🎉 

@rjmholt If you want to try it here, all you gotta do is add `<LangVersion>Preview</LangVersion>`

Reply:
From what I've read, the only drawback I know of with default interfaces is that they won't support invoking them via `base.` in this release due to the complexity of the implementation. That's been pushed to v.Next, however, I don't think that will get in the way of what you're doing here.